### PR TITLE
feat: expose merge strategy for configuration when automerging

### LINKED ISSUE #7184: Make merge strategy and message configurable for Bitbucket
**What would you like Renovate to be able to do?**

I would like to configure the merge strategy and message when renovate auto merges pull requests, because the commit history is unnecessarily inflated:

e.g.
``` 
auto merged
chore(deps): update dependency aws-sdk to v2.746.0
auto merged
auto merged
chore(deps): update aws-cdk monorepo to v1.62.0
chore(deps): update dependency @types/node to v14.6.4
auto merged
chore(deps): update dependency @types/jest to v26.0.13
auto merged
chore(deps): update dependency aws-sdk to v2.744.0
```

**Did you already have any implementation ideas?**

Use configuration here https://github.com/renovatebot/renovate/blob/012561f69afe0baebc689b045df7a1c65b31a567/lib/platform/bitbucket/index.ts#L743-L745

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

No

**Is this a feature you'd be interested in implementing yourself?**

Currently not possible

--

I remember reading a similar issue that was closed, please link if you find

### PR DESCRIPTION
**Pull request by jbirch-atlassian**
## Changes:

Expose the underlying merge strategy used by our platforms as a
configurable option for each repository to configure with the option
`automergeStrategy`. Additionally implement `automergeStrategy` for the
Bitbucket Cloud platform, as the only platform that doesn't currently
respect the remote repository's configured merge strategy.

## Context:

closes #7184 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by jbirch-atlassian** [comment]
I have two areas of feedback here that I'm happy to either modify this Pull Request for, or submit in a follow-up Pull Request.

1. From an internal ergonomics perspective, I think I'd like `mergeStrategy` to always be required on `mergePr`.
2. From a user ergonomics perspective, I think I'd like to extract the merge strategy into some form of exported function on each platform like `isMergable(pr, strat)` or something, so that it's possible to do a dry-run and be told "Renovate will do nothing, becayse platform X does not support strategy Y". At the moment, dry-run will emit what Renovate is *instructed* to do, and could possibly be told at runtime "It turns out platform X can't do Y".

**Comment by jbirch-atlassian** [comment]
image

The screenshot here, from bottom to top, shows:
* A pull request where the repository default is merge commit, and `automergeStrategy` was set to `auto`. It resulted in a merge commit.
* A pull request where the repository default is merge commit, and `automerge Strategy` was set to `squash`. It resulted in a squashed commit.
* A pull request where the repository default is merge commit, and `automerge Strategy` was set to `rebase` (unsupported by Bitbucket). It resulted in no changes between `1630639` and `3ef5842`.
* A pull request where the repository default is squash, and `automerge Strategy` was set to `auto`. It resulted in a squashed commit.

**Comment by jbirch-atlassian** [comment]
I was rude with force-pushes because I noticed an error immediately after pushing, before anyone had a chance to review. The error was an incomplete list of `allowedValues` in lib/config/definitions.ts.

**Comment by jbirch-atlassian** [comment]
One final point of discussion: this PR is a rather in depth approach to the original ticket, with the goal of making the strategy changable on a per-renovate-configuration level, as per the discussion in #7184. However, the only *strictly* required change in order to resolve #7184 is just... Don't provide the optional `merge_strategy` parameter to the Bitbucket Cloud API.

If we don't have the appetite yet to take on exposing the merge strategy as a configuration item, I'm happy to put up those changes as a less ambitious PR instead.

Thanks again for your time and advice, y'all.

File: lib/platform/bitbucket/index.ts
Comment (viceice) [inline lib/platform/bitbucket/index.ts:4]:
```suggestion
import type { MergeStrategy } from '../../config/types';
```

File: lib/platform/bitbucket/utils.ts
Comment (viceice) [inline lib/platform/bitbucket/utils.ts:2]:
```suggestion
import type { MergeStrategy } from '../../config/types';
```

File: lib/platform/bitbucket/utils.ts
Comment (viceice) [inline lib/platform/bitbucket/utils.ts:60]: Please move new  type and interface definitions to `./types.ts` and do a `import type`

**Review by viceice — COMMENTED** [review]
More later

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:195]:
```suggestion
## automergeStrategy

This setting is only applicable if you opt-in by configuring `automerge` to `true` and `automergeType` to `pr` for any of your dependencies.

The automerge strategy defaults to `auto`, in which Renovate will make its best guess as to how to merge pull requests.
This generally results in Renovate respecting the strategy configured in the platform itself for the repository if possible.
Acceptable values are:

- `auto`, in which the choice is left to Renovate
- `fast-forward`, which generally involves no new commits in the resultant tree, but "fast-forwarding" the main branch reference
- `merge-commit`, which generally involves synthesizing a new merge commit
- `rebase`, which generally involves rewriting history as part of the merge — but usually retaining the individual commits
- `squash`, which generally involves flattening the commits that are being merged into a single new commit

Not all platforms support all pull request merge strategies.
In cases where a merge strategy is not supported by the platform, Renovate will hold off on merging instead of silently merging in a way you didn't wish for.

The only platform that supports `automergeStrategy` is Bitbucket Cloud.

```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some initial changes, more to come later.

File: lib/platform/bitbucket/index.md
Comment (viceice) [inline lib/platform/bitbucket/index.md:6]:
```suggestion
- `automergeStrategy=rebase` is unsupported [bcloud-16610] by Bitbucket Cloud.
```
Typo?

File: lib/platform/gitlab/index.md
Comment (viceice) [inline lib/platform/gitlab/index.md:1]:
```suggestion
# GitLab
```

File: lib/platform/types.ts
Comment (viceice) [inline lib/platform/types.ts:1]:
```suggestion
import type { MergeStrategy } from '../config/types';
```

File: lib/platform/types.ts
Comment (viceice) [inline lib/platform/types.ts:159]: @rarkins i think we should refactor this to object arg after this pr

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/platform/bitbucket/index.md
Comment (jbirch-atlassian) [reply lib/platform/bitbucket/index.md:6]: This was a deliberate choice and not a typo, such that the text `is unsupported` is a link itself.

The initial commit displays like:
image

The suggested change displays like:
image

Which is more confusing to read in my opinion — it's not really a sentence. If the style of documents here is to make hyperlinks overt and explicit instead of hyperlinks in existing text, I can change this into something like the following?

```
- `automergeStrategy=rebase` is unsupported by Bitbucket Cloud (See: [BCLOUD-16610][bcloud-16610])
```

In light of this, would you prefer the intended commit, the suggested change, or the third option presented above?

File: lib/platform/gitlab/index.md
Comment (jbirch-atlassian) [reply lib/platform/gitlab/index.md:1]: This was chosen to align with the platform documentation at https://docs.renovatebot.com/modules/platform/ (IE, documentation of the platform), and not the name of the module this is in (IE, documentation of the module). Happy to take this change on, but please check the titles on lib/platform/github/index.md and lib/platform/azure/index.md and see if you'd like them similarly renamed for consistency of documentation.

File: lib/platform/bitbucket/utils.ts
Comment (jbirch-atlassian) [reply lib/platform/bitbucket/utils.ts:60]: `lib/platform/bitbucket/types.ts` does not yet exist. Would you prefer me to move *all* type and interface definitions in this file into that new file, or just the ones added here?

File: lib/platform/bitbucket-server/index.md
Comment (HonkingGoose) [inline lib/platform/bitbucket-server/index.md:13]:
```suggestion
- The `automergeStrategy` configuration option has not been implemented for this platform, and all values behave as if the value `auto` was used. Renovate will implicitly use the merge strategy configured as 'default' in the Bitbucket Server repository itself, and this cannot be overridden yet
```

We use one sentence per line pattern in our `*.md` files. 😉
We also drop the final `.` on a ordered/unordered list.

File: lib/platform/bitbucket/index.md
Comment (HonkingGoose) [reply lib/platform/bitbucket/index.md:6]:
```suggestion
- `automergeStrategy=rebase` not supported by BitBucket Cloud, see [Jira issue BCLOUD-16610](https://jira.atlassian.com/browse/BCLOUD-16610)
```

It should always be clear what you'll get when you click on a link. So, I'll take: "modified version of third option". See my suggestion in this comment. 😉 

I don't know if that `[BCLOUD-16610][bcloud-16610])` link style will work on our final page, so I'm taking the precaution to use a fully qualified URL here. 😄

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:187]: The way Renovate rebases **is not the same** as the way Git rebases.

Git rebases by making new commits and putting those on top of the rebased branch.

Renovate "rebases" by just doing a new run and force pushes a single new commit.
So it does not "retain individual commits". 😉 

```suggestion
- `rebase`, Renovate starts a fresh run and pushes a single new commit, that's up-to-date with the target branch
```

Maybe something like this is better? Not 100% sure of the correct wording. 😄

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: lib/platform/azure/index.md
Comment (JamieMagee) [inline lib/platform/azure/index.md:5]: Can you create an issue for this instead? Otherwise I'm never going to be able to find this, and it'll never get implemented.

Comment (jbirch-atlassian) [reply lib/platform/azure/index.md:5]: Instead of, or in addition to?

Comment (JamieMagee) [reply lib/platform/azure/index.md:5]: You caught my edit 😄 I initially thought instead of, but I think in addition to would be better. Thank you!

File: docs/usage/configuration-options.md
Comment (jbirch-atlassian) [reply docs/usage/configuration-options.md:187]: Hm yes, I see. This `rebase` is outside of renovate's usually rebase behaviour, and I added it here because most of the other platforms support something that looks like a merge rebase. I see two ways forward here:

1) Remove it. The only platform where this is exposed doesn't support a rebase-alike anyway.
2) Change it to `merge-rebase`, so as not to conflict with the usage of "rebase" that Renovate does when it does git directly.

I added it as a general thing because other platforms support it to various levels, but it might eventually become more complicated anyway — some platforms like Bitbucket Server support both `rebase + merge --no-ff` as well as `rebase + merge --ff-only` as explicit choices, and "rebase" alone isn't enough to describe this. The description I've added here attempted to be vague about these differences, committing only to "And there might be some history rewriting...". However, in these merge strategies, the commit contents do tend to keep existing, even if the history is rewritten.

Comment (HonkingGoose) [reply docs/usage/configuration-options.md:187]: This is going over my head, I know and use `git rebase -i upstream/main` to fixup/squash/drop commits, on my feature branches to clean up my work before I push it. But I'm not familiar with all the ways in which different platforms handle rebasing.

I'll defer to @rarkins's judgement on the wording here, as I'm not qualified to choose here. 😉

Comment (HonkingGoose) [reply docs/usage/configuration-options.md:187]: Oh by the way, we do have an issue (https://github.com/renovatebot/renovate/issues/9020) to replace the term `rebase` with `refresh`. Maybe the word `refresh` could be useful here as well?

Comment (jbirch-atlassian) [reply docs/usage/configuration-options.md:187]: Again, it's not our (IE, Renovate's) `rebase` terminology here, it's the `rebase` as referred to by the platforms that are hosting the repository. For example:

* Bitbucket Server has "rebase" strategies outlined in https://confluence.atlassian.com/bitbucketserver/pull-request-merge-strategies-844499235.html#Pullrequestmergestrategies-Mergestrategies.
* Github has a "rebase" strategy outlined in https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#rebase-and-merge-your-pull-request-commits
* Gitea has a "rebase" strategy added in https://github.com/go-gitea/gitea/pull/3188

None of these platforms refer to these merge strategies as "refresh strategies". The use of "refresh" inside Renovate to refer to the process by which Renovate keeps branches up to date is nice, but completely orthogonal to how "rebase" is used here.

Comment (jbirch-atlassian) [reply docs/usage/configuration-options.md:187]: @rarkins, Could I get you to have a read through of this comment chain specifically and answer @HonkingGoose and mine's discussion?

File: lib/platform/bitbucket/utils.ts
Comment (jbirch-atlassian) [reply lib/platform/bitbucket/utils.ts:60]: @viceice, could you elaborate on if you'd prefer:

* Create this new file, add only my stuff in it (IE: Keep the PR small)
* Create this new file, add all the appropriate stuff into it (IE: Help the code conform to style)

File: lib/platform/gitlab/index.md
Comment (jbirch-atlassian) [reply lib/platform/gitlab/index.md:1]: @viceice, It's safe to assume that the other files are titled appropriately?

File: lib/platform/bitbucket/utils.ts
Comment (viceice) [reply lib/platform/bitbucket/utils.ts:60]: for this pr only move new types and interfaces. Also prefer interfaces if possible.

File: docs/usage/configuration-options.md
Comment (viceice) [reply docs/usage/configuration-options.md:187]: @HonkingGoose We should leave as is, as this `renbase` means which option the platform uses to merge the pr. It has nothing to do with the renovate way of recreate fresh commit and force-push.

@jbirch-atlassian I think we can leave as is.

File: lib/platform/gitlab/index.md
Comment (viceice) [reply lib/platform/gitlab/index.md:1]: Yes, please use my suggestion for now.

File: lib/platform/types.ts
Comment (viceice) [reply lib/platform/types.ts:159]: #10830

File: docs/usage/configuration-options.md
Comment (HonkingGoose) [reply docs/usage/configuration-options.md:187]: Oh, if we're talking about the platforms `rebase` strategy, then we can keep things as-is. 😉

File: lib/platform/azure/index.md
Comment (jbirch-atlassian) [reply lib/platform/azure/index.md:5]: This has been formally requested in #10851

File: lib/platform/bitbucket/utils.ts
Comment (jbirch-atlassian) [reply lib/platform/bitbucket/utils.ts]: Resolved in a4e10f9

**Review by viceice — COMMENTED** [review]
Please open an issue for all platforms missing implementation too.

**Comment by jbirch-atlassian** [comment]
@viceice: I'm not sure what's happening, but you're listed as "requesting changes" despite everything being resolved. I *think* it's coming from the changes in `lib/platform/bitbucket/index.md` being resolved with the suggestions in a comment, not with your suggestion.

Is everything above board, or am I just failing to GitHub today?

**Comment by viceice** [comment]
The request changes will resolve when I approve this pr.

Please open issues for the platforms which are missing this feature in this PR. After that I'll approve. 😉

File: lib/config/definitions.ts
Comment (HonkingGoose) [inline lib/config/definitions.ts:1290]:
```suggestion
      'The merge strategy to use when automerging PRs. Used only if <code>automergeType=pr</code> or <code>automergeType=pr-comment</code>',
```

@viceice I was thinking maybe we could do this? I don't know if this is valid code though for this file?

File: lib/platform/azure/index.md
Comment (HonkingGoose) [inline lib/platform/azure/index.md:7]:
```suggestion
- The `automergeStrategy` configuration option has not been implemented for this platform, and all values behave as if  the value `auto` was used. Renovate will use the merge strategy configured in the Azure Repos repository itself, and this cannot be overridden yet.
```

This should be on a single line. 😉

File: lib/platform/gitea/index.md
Comment (HonkingGoose) [inline lib/platform/gitea/index.md:9]: Put this on on a single line as well.

File: lib/platform/github/index.md
Comment (HonkingGoose) [inline lib/platform/github/index.md:5]: Put this on on a single line as well.

File: lib/platform/gitlab/index.md
Comment (HonkingGoose) [inline lib/platform/gitlab/index.md:5]: Put this on on a single line as well.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some things I noticed just now.

One thing is for @viceice to apply if OK, the others you can do yourself. 😉

**Comment by jbirch-atlassian** [comment]
I guess it's just I didn't know how to GitHub! Thanks for the quick education, appreciate it.

#10867, #10868, #10869, and #10870 have been created for the remaining platforms.

File: lib/config/definitions.ts
Comment (jbirch-atlassian) [reply lib/config/definitions.ts:1290]: Ah, I'll need to change this anyway — `automergeType=pr-comment` doesn't result in a merge happening (And so the merge strategy doesn't apply). I'll leave this open for the approval or not of the `code` tags, but I'll need to resolve anyway by removing the automergeType=pr-comment bit.

Comment (viceice) [reply lib/config/definitions.ts:1290]:
```suggestion
      'The merge strategy to use when automerging PRs. Used only if `automergeType=pr`',
```

Should work too. 🙃

**Review by viceice — COMMENTED** [review]
So the last small thing before we are haape to merge.

File: lib/config/definitions.ts
Comment (HonkingGoose) [inline lib/config/definitions.ts:1290]:
```suggestion
      'The merge strategy to use when automerging PRs. Used only if `automergeType=pr`.',
```

I think the style here is that we use a full stop at the end of the sentence. 😉

File: lib/platform/azure/index.md
Comment (HonkingGoose) [inline lib/platform/azure/index.md:5]:
```suggestion
- The `automergeStrategy` configuration option has not been implemented for this platform, and all values behave as if the value `auto` was used. Renovate will use the merge strategy configured in the Azure DevOps / Azure DevOps Server repository itself, and this cannot be overridden yet
```

Are we sure it's `Azure Repos` and not something else?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
2 small things, looks good to me otherwise. 😉

File: lib/platform/azure/index.md
Comment (jbirch-atlassian) [reply lib/platform/azure/index.md:5]: Strictly speaking, yes. Azure Repos is the product included in the Azure DevOps suite of services (In the same way that Azure Boards is "Microsoft Jira", for instance). The product pages for [Azure DevOps](https://azure.microsoft.com/en-us/services/devops/) and [Azure DevOps Server](https://azure.microsoft.com/en-us/services/devops/server/) are what I'm basing this on.

However, if we've established a convention of referring to Azure Repos as "Azure DevOps" for the purposes of Renovate (which is based around repositories), I can appreciate if we'd like to make this change.

Comment (HonkingGoose) [reply lib/platform/azure/index.md:5]: Let's be technically correct then, and use Azure Repos as name. 😉

**Review by HonkingGoose — DISMISSED** [review]

**Comment by jbirch-atlassian** [comment]
Hey y'all, I think this one should be good to go now. I'll swing back around in a week or so and poke folk again if there's no movement.

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**
